### PR TITLE
[AutoDiff] switch array and floating point derivatives to @derivative(of:)

### PR DESCRIPTION
**Pull request by marcrasi**
Also I added a test for the `FloatingPoint.swift` derivatives because there isn't an existing test for them.

Once this and https://github.com/apple/swift/pull/28933 are merged, there will be no more vjp:s in the stdlib.

### CODE REVIEW

**Review by rxwei — APPROVED** [review]
Nice!

Minor comment: The title initially came across as "removing VJPs from stdlib", and I only realized it's removing `@differentiable(vjp: ...)`s when I looked at the diff. How about rewording the PR title to something like "[AutoDiff] Switch to '@derivative(of:)' for array and floating point derivative registrations"?

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Review by dan-zheng — COMMENTED** [review]
Is it (or will it eventually be) necessary to keep `@differentiable` attributes on the `public` original functions so that the derivatives are registered with `public` access?

---

We discussed this as part of derivative visibility. Currently, derivative visibility is equal to the access level of the `@derivative` declaration, but we could potentially decouple it (e.g. by adding syntax like `@derivative(of: foo, access: public)`).

**Comment by rxwei** [comment]
We discussed this and preferred making the derivative registration have the same visibility as the derivative function. The only thing that's preventing this is the fact that we do not yet support anonymous function declarations (with ` func _`) and do not want to export named functions as public symbols.

The current code pattern (as it is in the library) will likely not compile in the final design, because when `@differentiable` is put on a public declaration, it cannot be satisfied with an internal `@derivative(of:)` registration.

File: test/AutoDiff/floating_point.swift.gyb [hunk 1]
```diff
@@ -0,0 +1,36 @@
+// RUN: %empty-directory(%t)
+// RUN: %gyb %s -o %t/floating_point.swift
+// RUN: %target-build-swift %t/floating_point.swift -o %t/a.out
+// RUN: %target-run %t/a.out
```
Comment (dan-zheng) [inline test/AutoDiff/floating_point.swift.gyb]: I wonder if `%target-run-simple-swiftgyb` works? It's a one-liner.
```suggestion
RUN: %target-run-simple-swiftgyb 
```

Comment (marcrasi) [reply test/AutoDiff/floating_point.swift.gyb]: Yes, this works. Done!

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow

**Comment by marcrasi** [comment]
@swift-ci please test tensorflow
